### PR TITLE
 KFSPTS-17966 add invoice number and date to DV in create accounting document job

### PR DESCRIPTION
**Pull request by jhulslander**

### CODE REVIEW

File: src/test/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImplTest.java [hunk 1]
```diff
@@ -677,6 +688,10 @@ private void assertCuDisbursementVoucherDocumentsCorrect(CuDisbursementVoucherDo
                 actualDvDocument.getDvNonEmployeeTravel().getDvNonEmployeeExpenses(), this::assertTravelExpenseCorrect);
         assertObjectListIsCorrect("non employee prepaid travel expenses aren't correct", expectedDvDocument.getDvNonEmployeeTravel().getDvPrePaidEmployeeExpenses(), 
                 actualDvDocument.getDvNonEmployeeTravel().getDvPrePaidEmployeeExpenses(), this::assertTravelExpenseCorrect);
+        assertEquals("Due Dates should match", dateFormat.format(actualDvDocument.getDisbursementVoucherDueDate()), 
+                dateFormat.format(expectedDvDocument.getDisbursementVoucherDueDate()));
+        assertEquals("Invoice Dates should match", actualDvDocument.getInvoiceDate(), expectedDvDocument.getInvoiceDate());
+        assertEquals("Invoice numbers should match", actualDvDocument.getInvoiceDate(), expectedDvDocument.getInvoiceDate());
```
Comment (cah292) [inline src/test/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImplTest.java]: Is it safe to assume that the disbursementVoucherDueDate is non-null when it's passed to the date formatter? I think that's the case based on the default-due-date feature, but wanted to make sure.

Also, when calling assertEquals(), the arg for the expected value should precede the one for the actual value.

File: src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/AccountingXmlDocumentEntryFixture.java [hunk 2]
```diff
@@ -459,6 +459,37 @@
                     AccountingXmlDocumentBackupLinkFixture.CORNELL_INDEX_PAGE,
                     AccountingXmlDocumentBackupLinkFixture.DFA_INDEX_PAGE),
             CuDisbursementVoucherDocumentFixture.JANE_DOE_DV_DETAIL),
+    DV_DOC_TEST1(1, CuFPTestConstants.DISBURSEMENT_VOUCHER_DOC_TYPE,
```
Comment (cah292) [inline src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/AccountingXmlDocumentEntryFixture.java]: The new fixtures were placed right in between some other ones that go together for a separate test case. Can this new fixture be moved to a different position, to make it clear that it should not be grouped with the "MULTI_DOC_TYPE_TEST..." fixtures?

File: src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/AccountingXmlDocumentEntryFixture.java [hunk 3]
```diff
@@ -459,6 +459,37 @@
...
+            adHocRecipients(
+                    AccountingXmlDocumentAdHocRecipientFixture.JDH34_APPROVE,
+                    AccountingXmlDocumentAdHocRecipientFixture.SE12_FYI,
+                    AccountingXmlDocumentAdHocRecipientFixture.CCS1_COMPLETE,
+                    AccountingXmlDocumentAdHocRecipientFixture.NKK4_ACKNOWLEDGE),
+            backupLinks(
+                    AccountingXmlDocumentBackupLinkFixture.CORNELL_INDEX_PAGE,
+                    AccountingXmlDocumentBackupLinkFixture.DFA_INDEX_PAGE),
+            CuDisbursementVoucherDocumentFixture.JANE_DOE_DV_DETAIL),
+    DV_DOC_TEST2(2, CuFPTestConstants.DISBURSEMENT_VOUCHER_DOC_TYPE,
```
Comment (cah292) [inline src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/AccountingXmlDocumentEntryFixture.java]: This fixture should also be moved to wherever the "DV_DOC_TEST1" fixture gets moved to.

File: src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java [hunk 4]
```diff
@@ -1,18 +1,26 @@
...
 
 public enum CuDisbursementVoucherDocumentFixture {
     EMPTY(),
-    JANE_DOE_DV_DETAIL("DISB", "Doe, Jane", "X", "E", 50, "Freeville", "Jane Doe", 25,
+    JANE_DOE_DV_DETAIL("DISB", "Doe, Jane", "X", "E", 50, "Freeville", "Jane Doe", 25, "7/24/2018", StringUtils.EMPTY, StringUtils.EMPTY,
             buildExpenseFixtureArray(CuDisbursementVoucherDocumentNonEmployeeTravelExpenseFixture.DELTA,
                     CuDisbursementVoucherDocumentNonEmployeeTravelExpenseFixture.OTHER_LODGING),
+            buildExpenseFixtureArray(CuDisbursementVoucherDocumentNonEmployeeTravelExpenseFixture.PREPAID_AVIS,
+                    CuDisbursementVoucherDocumentNonEmployeeTravelExpenseFixture.PREPAID_OTHER)),
+    JOHN_DOE_DV_DETAIL("DISB", "Doe, John", "X", "E", 50, "Freeville", "Jane Doe", 25, StringUtils.EMPTY, "04/15/2020", "12321",
```
Comment (cah292) [inline src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java]: Should the second of the two name strings be "John Doe" instead?

File: src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java [hunk 5]
```diff
@@ -24,6 +32,9 @@
     public final String conferenceDestination;
     public final String nonEmployeeTravelerName;
     public final Integer nonEmployeeCarMileage;
+    public final Date dueDate;
+    public final Date invoiceDate;
```
Comment (cah292) [inline src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java]: Date objects are mutable. If you want this fixture enum to remain immutable, you could create some getters that generate a new Date object when invoked, and within the fixture you could store the date information in some other immutable form (like leaving it as a String or using a Joda DateTime object).

File: src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java [hunk 6]
```diff
@@ -53,6 +69,23 @@ private CuDisbursementVoucherDocumentFixture(String bankCode, String contactName
...
+        if (StringUtils.isNotEmpty(invoiceDateString)) {
+            this.invoiceDate = new Date(StringToJavaDateAdapter.parseToDateTime(invoiceDateString).getMillis());
+        } else {
+            this.invoiceDate = null;
+        }
+    }
+    
+    private static Date calculateDefaultDueDate() {
+        Calendar cal = Calendar.getInstance();
+        cal.add(Calendar.DATE, 1);;
```
Comment (cah292) [inline src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java]: Unneeded extra semicolon.

File: src/test/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImplTest.java (hunk 1 above)
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImplTest.java]: Yes, the due should never be null, I shall switch the order

File: src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/AccountingXmlDocumentEntryFixture.java (hunk 2 above)
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/AccountingXmlDocumentEntryFixture.java]: Yes, I shall move them down

File: src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/AccountingXmlDocumentEntryFixture.java (hunk 3 above)
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/AccountingXmlDocumentEntryFixture.java]: moved

File: src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java (hunk 4 above)
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java]: yes it should, I shall change

File: src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java (hunk 6 above)
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java]: I shall remove

File: src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java (hunk 5 above)
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java]: switched to joda datetime

**Comment by jhulslander** [comment]
I've made changes as noted, this is ready for review again

File: src/test/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImplTest.java [hunk 7]
```diff
@@ -677,6 +688,10 @@ private void assertCuDisbursementVoucherDocumentsCorrect(CuDisbursementVoucherDo
                 actualDvDocument.getDvNonEmployeeTravel().getDvNonEmployeeExpenses(), this::assertTravelExpenseCorrect);
         assertObjectListIsCorrect("non employee prepaid travel expenses aren't correct", expectedDvDocument.getDvNonEmployeeTravel().getDvPrePaidEmployeeExpenses(), 
                 actualDvDocument.getDvNonEmployeeTravel().getDvPrePaidEmployeeExpenses(), this::assertTravelExpenseCorrect);
+        assertEquals("Due Dates should match", dateFormat.format(expectedDvDocument.getDisbursementVoucherDueDate()), 
+                dateFormat.format(actualDvDocument.getDisbursementVoucherDueDate()) );
+        assertEquals("Invoice Dates should match", actualDvDocument.getInvoiceDate(), expectedDvDocument.getInvoiceDate());
+        assertEquals("Invoice numbers should match", actualDvDocument.getInvoiceDate(), expectedDvDocument.getInvoiceDate());
```
Comment (cah292) [inline src/test/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImplTest.java]: These two assertions still need to adjust the positions of the expected-value and actual-value args. Sorry if that wasn't clear from my initial comment.

Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImplTest.java]: no worries, I think I glossed over them, I fixed

**Comment by jhulslander** [comment]
I made further changes, this is ready for review again

**Pull request merged**